extensions/ql-vscode/src/codeql-task-provider.ts

```diff
@@ -0,0 +1,111 @@
+  }
+
+  public resolveTask(_task: vscode.Task): vscode.Task | undefined {
+    const command: string = _task.definition.command;
+    if (command) {
+      const definition: CodeQLTaskDefinition = <any>_task.definition;
+      return this.getTask(definition.command, definition.commandArgs ?? [], definition);
+    }
+    return undefined;
+  }
+
+  private getTasks(): vscode.Task[] {
+    if (this.tasks !== undefined) {
+      return this.tasks;
+    }
+    // Hard-code examples for now. Not sure if we can do this in a better way.
+    const commands: string[] = [
+      'resolve languages',
+      'version',
+      'pack install',      // TODO: Specify which qlpack to install dependencies for.
+      'resolve queries',   // TODO: Specify which query, directory, or suite to resolve. Also specify search path/additional packs.
+      'resolve qlpacks',   // TODO: Specify a search path/additional packs.
+      // 'pack publish',     // Needs a PAT (from octokit?)
+      // 'pack download',    // Needs a PAT (from octokit?)
+      // 'database create'   // Needs additional args.
+    ];
+    const commandArgs: string[][] = [[]];
+
+    const tasks: vscode.Task[] | undefined = [];
+    commands.forEach(command => {
+      commandArgs.forEach(commandArgsGroup => {
+        tasks.push(this.getTask(command, commandArgsGroup));
+      });
+    });
+    this.tasks = tasks;
+    return this.tasks;
+  }
+
+  private getTask(command: string, commandArgs: string[], definition?: CodeQLTaskDefinition): vscode.Task {
+    if (definition === undefined) {
+      definition = {
+        type: CodeQLTaskProvider.CodeQLType,
+        command,
+        commandArgs
+      };
+    }
+    return new vscode.Task(definition, vscode.TaskScope.Workspace, `${command} ${commandArgs.join(' ')}`,
+      CodeQLTaskProvider.CodeQLType, new vscode.CustomExecution(async (): Promise<vscode.Pseudoterminal> => {
+        return new CodeQLTaskTerminal(this.cliServer, command, commandArgs);
+      }));
+  }
+}
+
+class CodeQLTaskTerminal implements vscode.Pseudoterminal {
+  private writeEmitter = new vscode.EventEmitter<string>();
+  onDidWrite: vscode.Event<string> = this.writeEmitter.event;
+  private closeEmitter = new vscode.EventEmitter<number>();
+  onDidClose?: vscode.Event<number> = this.closeEmitter.event;
+
+  private fileWatcher: vscode.FileSystemWatcher | undefined;
+
+  // TODO: Open the terminal (i.e. run the task) in an appropriate folder (e.g. the enclosing folder of the currently active editor).
+  // Currently, it is run from the first workspace folder (codeql-custom-queries-cpp in the starter workspace).
+  constructor(private cliServer: CodeQLCliServer, private command: string, private commandArgs: string[]) {
+  }
+
+  async open(): Promise<void> {
+    await this.runCli([this.command], this.commandArgs);
+  }
+
+  close(): void {
+    // The terminal has been closed. Shutdown the build.
+    if (this.fileWatcher) {
+      this.fileWatcher.dispose();
+    }
+  }
```

Comment:
I think you can make this class a `DisposableObject`. Then when you first create the file watcher, you can just do `this.push(this.fileWatcher)`.

This close method will not be necessary.

Reply:
If `close`  is an abstract super method, then you can change the body to `this.dispose();`.